subsys/bluetooth/shell/bt.c

```diff
@@ -697,6 +697,66 @@ static int cmd_name(const struct shell *sh, size_t argc, char *argv[])
 	return 0;
 }
 
+/**
+ * @brief Read a hexadecimal number like "0x12aB" or "12aB" into 0x12aB.
+
+ * @retval 0 Success.
+ * @retval -EINVAL Input is either not a hexadecimal number or too long to fit in uint16_t.
+ */
+static int hexstr2uint16(size_t hexlen, char hex[hexlen], uint16_t *output)
+{
+	/* Ignore 0x at start. */
+	if (hexlen >= 2 && hex[0] == '0' && hex[1] == 'x') {
+		hex = &hex[2];
+		hexlen -= 2;
+	}
+
+	if (hexlen == 0 || hexlen > 4) {
+		return -EINVAL;
+	}
+
+	*output = 0;
+	for (size_t i = 0; i < hexlen; i++) {
+		uint8_t digit;
+		int err = char2hex(hex[i], &digit);
+
+		if (err) {
+			return -EINVAL;
+		}
+
+		*output = (*output) * 0x10 + digit;
+	}
+
+	return 0;
+}
+
+static int cmd_appearance(const struct shell *sh, size_t argc, char *argv[])
+{
+	if (argc == 1) {
+		shell_print(sh, "Bluetooth Appearance: 0x%04x", bt_get_appearance());
+		return 0;
+	}
+
+	if (argc == 2) {
+		int err;
+		uint16_t new_appearance;
+
+		err = hexstr2uint16(strlen(argv[1]), argv[1], &new_appearance);
```

Comment:
Besides the special handling of "0x" in the beginning, wouldn't it have been simpler to reuse `hex2bin()` instead of creating a new function? It'd then be (excluding again the "0x" handling):
```
if (hex2bin(argv[1], strlen(argv[1]), &new_apperance, sizeof(new_appearance)) != sizeof(new_appearance) {
        shell_error(sh, "Could not parse argument.");
        return -EINVAL;
}
```

I think it would be justified to add "0x" prefix handling to `hex2bin()` so that could be its own patch.

Reply:
The functions are subtly different. `hexstr2uint16` parses a human-readable uint16. `hex2bin` reads an array of uint8. I think it may need a better name. Any suggestions?

Reply:
Wouldn't the end result still be the same? Only thing you'll need to do is a `sys_le16_to_cpu()` conversion since the resulting binary buffer would be encoded in little-endian format.

Reply:
Another reason I'm pushing against this is that I don't think we should be doing custom helpers like this and make them private to a single shell module. If there's really value having a uint16_t-specific version (which I'm not convinced of, and then you should ask if uint32_t etc is also justified) then it should be made available through the same header file that provides `hex2bin()` etc.

Reply:
Ok. I'll use `hex2bin` and `sys_be16_to_cpu`.

Tangent question: I have a slight aversion to juggling bytes inside a number like `sys_be16_to_cpu` does. It bypasses the type system, so it's dangerous. It's easy to mistakenly do arithmetic on a jumbled uint16_t. Or to pass it somewhere that expects a native uint16_t. A proper type for big endian could be `struct uint16_be { uint8_t serialized[2] };`. I assume there is a practical reason for why this is not how we do it? Does it optimize worse? Are structs forbidden to be passed to functions in registers?

Reply:
I would welcome Zephyr to have byte-order specific integer types the way Linux has, so that we catch missing conversions and incompatible types already at the compilation phase. However, this is a much bigger discussion that shouldn't be had inside a specific PR like this. Currently the basic rule that I follow for that stack is this: if a fixed-size integer (uint8_t, uint16_t uint32_t, etc) is inside a packed struct, then it's assumed to have protocol-specific endianness, otherwise it's assumed to have host endianness. Of course the case with this PR is different (just command line parsing), but I'd claim it's a pretty rare scenario for a protocol stack.